### Sujal Choudhari         
#### 60003220216
### IT3 (2)                 
#### Python Programming Lab (4)

Questions for Practical Assignment Submission

- WA function to calculate arithmetic mean of variable number of values 

- WA lambda function to find largest of 2 numbers 

- Use lambda function to return a square value of a given number. 

- Write a program using filter() to filter out even numbers from a list. 

- Write a program using map() to filter out even numbers from a list. 

- Write a program using reduce() with a lambda expression to find product of elements in a list. 

Questions for Write-up

1. Describe how lambda functions differ from regular functions in terms of syntax.
2. Need for Lamda and higher order functions 
3. What is the purpose of Python's map function?
4. How is the reduce function different from map and filter?

### 1. WA function to calculate arithmetic mean of variable number of values 

In [17]:
import functools

def mean ( *args):
    sum = functools.reduce(lambda x,y: x + y,args)
    print(sum / len(args))

In [18]:
mean(5,6,7,8,65,4,43,43,4,4,4,5,545,4)

53.357142857142854


### 2. WA lambda function to find largest of 2 numbers 

In [19]:
x = lambda x,y: x if x > y else y
x(5,6)

6

### 3. Use lambda function to return a square value of a given number. 


In [20]:
b = lambda x: x ** 2
b(5)

25

### 4. Write a program using filter() to filter out even numbers from a list. 


In [21]:
lst = [5,6,7,4,3,32,2,2,0,2]

list(filter(lambda x: x % 2 ==0,lst))

[6, 4, 32, 2, 2, 0, 2]

### 5. Write a program using map() to filter out even numbers from a list. 


In [22]:
lst = [4,5,6,7,8,9,0,1,2,3]

# mapped = [a for a in list(map(lambda x: x if x % 2 == 0 else None,lst)) if a is not None]
# mapped = [a for a in list(map(lambda x: x if x % 2 == 0 else None,lst)) if a is not None]
mapped = [a for a in list(map(lambda x: x if x % 2 == 0 else None,lst)) if a is not None]

mapped

[4, 6, 8, 0, 2]

### 6. Write a program using reduce() with a lambda expression to find product of elements in a list. 

In [23]:
import functools
lst2 = [1,2,3,4,5,6,7,8,9]
functools.reduce(lambda x,y: x * y,lst2)

362880

# Questions


1. 
Write a Python program, which returns a list with 2-tuples. Each tuple consists of a the order number and the product of the price per items and the quantity. The product should be increased by 10,- € if the value of the order is smaller than 100,00 €. Write a Python program using lambda and map.

In [24]:
orders = [(1, 10, 20), (2, 500, 30), (3, 8, 15)]  # id, price, quantity
result = list(map(lambda x: (x[0], x[1] * x[2] if x[1] * x[2] >= 10000 else x[1] * x[2] + 10), orders))
print(result)

[(1, 210), (2, 15000), (3, 130)]


2. The same bookshop, but this time we work on a different list. The sublists of our lists look like this: [ordernumber, (article number, quantity, price per unit), ... (article number, quantity, price per unit) ] Write a program which returns a list of two tuples with (order number, total amount of order).

In [25]:
orders = [[1, (151, 2, 30), (152, 5, 15)], [2, (103, 10, 8), (104, 4, 5)]]
result = list(map(lambda x: (x[0], sum([i[1] * i[2] for i in x[1:]])), orders))
print(result) 

[(1, 135), (2, 100)]


3. Use python lambda and the map function to multiply every element of a list by two.

li = [5, 7, 22, 97, 54, 62, 77, 23, 73, 61]

output [10, 14, 44, 194, 108, 124, 154, 46, 146, 122]

In [26]:
li = [5, 7, 22, 97, 54, 62, 77, 23, 73, 61]
result = list(map(lambda x: x * 2, li))
print(result)

[10, 14, 44, 194, 108, 124, 154, 46, 146, 122]


4. Write a Python program to sort a list of tuples using Lambda. Original list of tuples: [('English', 88), ('Science', 90), ('Maths', 97), ('Social sciences', 82)] Sorting the List of Tuples: [('Social sciences', 82), ('English', 88), ('Science', 90), ('Maths', 97)]

In [27]:
og = [('English', 88), ('Science', 90), ('Maths', 97), ('Social sciences', 82)] 
sorted_list = sorted(og, key=lambda x: x[1])
sorted_list

[('Social sciences', 82), ('English', 88), ('Science', 90), ('Maths', 97)]

5. Write a Python program to extract year, month, date and time using Lambda. Sample Output: 2020-01-15 09:03:32.744178 2020 1 15 09:03:32.744178

In [28]:
# Program to extract year, month, date, and time using Lambda
import datetime
now = datetime.datetime.now()
extraction = lambda x: f"{x.date()} {x.time()}"
extraction(now)

'2024-02-22 11:42:31.906043'

6. Write a Python program to create the next bigger number by rearranging the digits of a given number. Original number: 12 Next bigger number: 21 Original number: 10 Next bigger number: False Original number: 201 Next bigger number: 210

In [29]:
def next_bigger_number(number):
    digits = [int(x) for x in str(number)]
    i = len(digits) - 2
    while i >= 0 and digits[i] >= digits[i + 1]:
        i -= 1
    
    if i == -1:
        return False
    
    j = len(digits) - 1
    while digits[j] <= digits[i]:
        j -= 1
    
    digits[i], digits[j] = digits[j], digits[i]
    
    digits[i + 1:] = reversed(digits[i + 1:])
    result = int(''.join(map(str, digits)))
    
    return result

In [30]:
og1 = 12
print(f"Given: {og1}\t Next bigger number: {next_bigger_number(og1)}")

og2 = 10
print(f"Given: {og2}\t Next bigger number: {next_bigger_number(og2)}")

og3 = 201
print(f"Given: {og3}\t Next bigger number: {next_bigger_number(og3)}")

Given: 12	 Next bigger number: 21
Given: 10	 Next bigger number: False
Given: 201	 Next bigger number: 210


7. Write a Python program to reverse strings in a given list of string values using lambda. 


Original lists: ['Red', 'Green', 'Blue', 'White', 'Black'] Reverse strings of the said given list: ['deR', 'neerG', 'eulB', 'etihW', 'kcalB']

In [31]:
original_list = ['Red', 'Green', 'Blue', 'White', 'Black']
reversed_list = list(map(lambda x: x[::-1], original_list))
print(reversed_list)

['deR', 'neerG', 'eulB', 'etihW', 'kcalB']


8. Write a Python program to count the occurrences of items in a given list using lambda.  
 Original list: [3, 4, 5, 8, 0, 3, 8, 5, 0, 3, 1, 5, 2, 3, 4, 2] Count the occurrences of the items in the said list: {3: 4, 4: 2, 5: 3, 8: 2, 0: 2, 1: 1, 2: 2}

In [32]:
original_list = [3, 4, 5, 8, 0, 3, 8, 5, 0, 3, 1, 5, 2, 3, 4, 2]

occurrences = dict(map(lambda x: (x, original_list.count(x)), original_list))

print(occurrences)

{3: 4, 4: 2, 5: 3, 8: 2, 0: 2, 1: 1, 2: 2}


9.Write a program using reduce() with a lambda expression to find product of elements in a list.

In [33]:
import functools
eles = [5,4,3,2,1] # 120

functools.reduce(lambda x,y : x *y,eles)

120

10. Write a Python program to remove all elements from a given list present in another list using lambda. Original lists: list1: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10] list2: [2, 4, 6, 8] Remove all elements from 'list1' present in 'list2: [1, 3, 5, 7, 9, 10]

In [34]:
oglist :list[int]= [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
remlist = [2, 4, 6, 8]

finallist = list(filter(lambda x: x not in remlist,oglist))
finallist

[1, 3, 5, 7, 9, 10]

11. Read the total marks of 5 students and convert it to the percentage.

    1. Filter the top 3 students percentage from above result

    2. In the list of numbers from 1 to 20, replace multiples of 3 with "Fizz", multiples of 5 with "Buzz", and multiples of both 3 and 5 with "FizzBuzz". Use the lambda function to get the task done

    3. Takes a list of numbers as input, computes the sum of the numbers, and prints the sum along with the average

In [35]:
# Read total marks for 5 students
marks = []
for i in range(5):
    mark = float(input(f"Enter the total marks for student {i + 1}: "))
    marks.append(mark)

# Convert marks to percentage
percentages = [(mark / 100) * 100 for mark in marks]

print("Percentages:", percentages)


Percentages: [10.0, 20.0, 66.0, 55.60000000000001, 434.0]


In [36]:
# Assuming percentages is already defined from the previous code
top_3_percentages = sorted(percentages, reverse=True)[:3]
print("Top 3:", top_3_percentages)


Top 3: [434.0, 66.0, 55.60000000000001]


In [37]:
numbers = list(range(1, 21))
result = list(map(lambda x: "Fizz"*(x % 3 == 0) + "Buzz"*(x % 5 == 0) or x, numbers))
print("FizzBuzz?:", result)


FizzBuzz?: [1, 2, 'Fizz', 4, 'Buzz', 'Fizz', 7, 8, 'Fizz', 'Buzz', 11, 'Fizz', 13, 14, 'FizzBuzz', 16, 17, 'Fizz', 19, 'Buzz']


In [38]:
def sum_and_average(number_list):
    total_sum = sum(number_list)
    average = total_sum / len(number_list)
    print(f"Sum: {total_sum}, Average: {average}")

input_numbers = [4, 8, 12, 16, 20]
sum_and_average(input_numbers)


Sum: 60, Average: 12.0
